compiler/hash-typecheck/src/ops/validate.rs

```diff
@@ -33,31 +35,136 @@ impl<'gs, 'ls, 'cd> AccessToStorageMut for Validator<'gs, 'ls, 'cd> {
+            term_ty_id,
+        };
+
+        let term = reader.get_term(simplified_term_id);
+        // @@PotentiallyIncomplete: there might are a few more checks we need to perform here, but
+        // this is not yet evident.
+        //
+        // @@Todo: actually implement these checks:
+        match term {
+            Term::Merge(_) => {
+                // Validate each inner term.
+                // Ensure all elements of the merge are not type functions, and are of the same level.
+                todo!()
+            }
+            Term::TyFn(_) => {
+                // Validate each member.
+                // Ensure the return value of each case is a subtype of the general return type.
+                todo!()
+            }
+            Term::Level1(level1_term) => match level1_term {
+                Level1Term::Tuple(_) => {
+                    // Validate each parameter
+                    todo!()
+                }
+                Level1Term::Fn(_) => {
+                    // Validate parameters and return type
+                    todo!()
+                }
+                Level1Term::ModDef(_) | Level1Term::NominalDef(_) => {
+                    // Nothing to do
+                    Ok(result)
+                }
+            },
+            Term::Level0(_) => {
+                // Validate the inner type and ensure it is level 1
+                todo!()
+            }
+            Term::Level2(_)
+            | Term::Level3(_)
+            | Term::Access(_)
+            | Term::Var(_)
+            | Term::AppSub(_)
+            | Term::TyFnTy(_)
+            | Term::AppTyFn(_)
+            | Term::Unresolved(_) => {
+                // Nothing to do, should have already been validated by the typer.
+                Ok(result)
+            }
+        }
+    }
+
+    /// Determine whether the given term is constant, i.e. has no side effects.
+    ///
+    /// This is the condition for the term to be able to be used as the return type of a type
+    /// function, or to be used in global scope.
+    pub fn term_is_constant(&mut self, term_id: TermId) -> TcResult<bool> {
+        // Make sure the term is valid first:
+        self.validate_term(term_id)?;
+        let simplified_term_id = self.simplifier().potentially_simplify_term(term_id)?;
+        let reader = self.reader();
+        let term = reader.get_term(simplified_term_id);
+        // @@Todo: actually implement this:
+        match term {
+            Term::Access(_) => todo!(),
+            Term::Var(_) => todo!(),
+            Term::Merge(_) => todo!(),
+            Term::TyFn(_) => todo!(),
+            Term::TyFnTy(_) => todo!(),
+            Term::AppTyFn(_) => todo!(),
+            Term::AppSub(_) => todo!(),
+            Term::Unresolved(_) => todo!(),
+            Term::Level3(_) => todo!(),
+            Term::Level2(_) => todo!(),
+            Term::Level1(_) => todo!(),
+            Term::Level0(_) => todo!(),
+        }
```

Comment:
How would you determine that a term is constant?

Reply:
Basically any expression that isn't a struct, type function, enum, trait, mod/impl, type function application, access, enum variant, or a function literal, is not constant.

Reply:
So what's a straight forward example of a non-constant term, I am guessing a runtime value?

Also, could this be useful for other applications when lowering?

Reply:
`land_on_the_moon()` is not constant. Neither is `3`. As far as the typechecker is concerned they are both runtime values. In the future this can be relaxed if literal types are implemented.

Reply:
This could be useful for lowering, but it would have to be a bit more sophisticated to actually be useful, like e.g. to detect pure functions Vs side effects.

Reply:
Yes I see now, this makes more sense (and it could definitely become more sophisticated when literal types are implemented).